muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingController.java

```diff
@@ -170,6 +171,26 @@ public WorkspaceAssessment updateWorkspaceAssessment(SchoolDataIdentifier worksp
 
   public void deleteWorkspaceAssessment(SchoolDataIdentifier workspaceIdentifier, SchoolDataIdentifier studentIdentifier, SchoolDataIdentifier workspaceAssesmentIdentifier) {
     gradingSchoolDataController.deleteWorkspaceAssessment(workspaceIdentifier, studentIdentifier, workspaceAssesmentIdentifier);
+    // #2716: When workspace assessment is removed, restore latest workspace assessment request
+    List<WorkspaceAssessmentRequest> requests = listWorkspaceAssessmentRequests(workspaceIdentifier.getDataSource(), workspaceIdentifier.getIdentifier(), studentIdentifier.getIdentifier());
+    requests.sort(new Comparator<WorkspaceAssessmentRequest>() {
+      public int compare(WorkspaceAssessmentRequest o1, WorkspaceAssessmentRequest o2) {
+        return o2.getDate().compareTo(o1.getDate()); // latest request first
+      }
+    });
+    // Update should cause the school data source to treat the request as active again (i.e. not handled)
+    WorkspaceAssessmentRequest latestRequest = requests.isEmpty() ? null : requests.get(0);
```

Comment:
By good coding practice, list methods should always return an empty list rather than null but it's admittedly better be safe than NPE.